mlflow/pyfunc/__init__.py

```diff
@@ -828,13 +828,158 @@ def _create_model_downloading_tmp_dir(should_use_nfs):
+def _convert_array_values(values, spark_type, spark_primitive_type_to_np_type):
+    """
+    Convert list or numpy array values to spark dataframe column values.
+    """
+    from pyspark.sql.types import ArrayType
+
+    # Get element type from a array<E> type or a array<array<E>> type.
+    # and get the array_dim value (1 or 2).
+    array_dim = 1
+    elem_type = spark_type.elementType
+    if isinstance(elem_type, ArrayType):
+        elem_type = elem_type.elementType
+        array_dim = 2
+
+    if type(elem_type) not in spark_primitive_type_to_np_type:
+        raise MlflowException(
+            "Unsupported array type field with element type "
+            f"{elem_type.simpleString()} in struct type.",
+            error_code=INVALID_PARAMETER_VALUE,
+        )
+    np_type = spark_primitive_type_to_np_type[type(elem_type)]
+
+    if array_dim == 1:
+        return [np.array(v, dtype=np_type) for v in values]
+    else:
+        return [list(np.array(v, dtype=np_type)) for v in values]
+
+
+def _get_spark_primitive_types():
+    from pyspark.sql import types
+
+    return (
+        types.IntegerType,
+        types.LongType,
+        types.FloatType,
+        types.DoubleType,
+        types.StringType,
+        types.BooleanType,
+    )
+
+
+def _check_udf_return_struct_type(struct_type):
+    from pyspark.sql.types import StructType, ArrayType
+
+    primitive_types = _get_spark_primitive_types()
+
+    for field in struct_type.fields:
+        field_type = field.dataType
+
+        if isinstance(field_type, primitive_types):
+            pass
+        elif isinstance(field_type, ArrayType):
+            if not _check_udf_return_array_type(field_type, allow_nested_struct=False):
+                return False
+        else:
+            return False
+
+    return True
+
+
+def _check_udf_return_array_type(array_type, allow_nested_struct):
+    from pyspark.sql.types import ArrayType, StructType
+
+    elem_type = array_type.elementType
+    primitive_types = _get_spark_primitive_types()
+
+    if isinstance(elem_type, primitive_types):
+        # 1D array
+        return True
+
+    if isinstance(elem_type, StructType):
+        if allow_nested_struct:
+            # Array of struct values.
+            return _check_udf_return_struct_type(elem_type)
+
+        return False
+
+    if isinstance(elem_type, ArrayType) and isinstance(elem_type.elementType, primitive_types):
+        # 2D array
+        return True
+
+    return False
+
+
+def _check_udf_return_type(data_type):
+    from pyspark.sql.types import ArrayType, StructType
+
+    primitive_types = _get_spark_primitive_types()
+    if isinstance(data_type, primitive_types):
+        return True
+
+    if isinstance(data_type, ArrayType):
+        return _check_udf_return_array_type(data_type, allow_nested_struct=True)
+
+    if isinstance(data_type, StructType):
+        return _check_udf_return_struct_type(data_type)
+
```

Comment:
What should be returned if we reach this line?

Reply:
Good catch. Updated.